### PR TITLE
docs: Add External Bears tutorial

### PR DESCRIPTION
**Pull request by Redridge**

### CODE REVIEW

File: docs/Users/Tutorials/External_Bears.rst [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+
+Welcome. This tutorial hopes to teach you how to use the
+``@external_bear_wrap`` decorator in order to write Bears in languages other
+than python.
+
+.. note::
+
+  This tutorial assumes that you already know the basics. If you are new please
+  refer to
+  :doc:`this link<Writing_Bears>`.
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: "this link" doesn't sound good. Just use "refer to the [Writing Bears](#link to the section) section below" maybe.

File: docs/Users/Tutorials/External_Bears.rst [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+
+  If you are planning to create a bear that uses an already existing tool (aka
+  linter), please refer to
+  :doc:`this link instead<Linter_Bears>`.
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: same here

File: docs/Users/Tutorials/External_Bears.rst [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+
+Why is This Useful?
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: Is this relevant to this tutorial? People looking to write a bear already know what coala is.

File: docs/Users/Tutorials/External_Bears.rst [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+-------------------
+
+coala is a great tool if you want to write static code analysis without having
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: Is this relevant to this tutorial? People looking to write a bear already know what coala is.

File: docs/Users/Tutorials/External_Bears.rst [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+Enabling developers to write external Bears means that they do not need to waste
+any time learning new languages and they can get faster to the interesting part,
+**static code analysis**.
+
+How Does This Work?
+-------------------
+
+By using the ``@external_bear_wrap`` decorator you will have all the necessary
+data sent to your external executable (filename, lines, settings) as a JSON
+string via stdin. Afterwards the analysis takes place in your executable that
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: stdin -> \`stdin\`

File: docs/Users/Tutorials/External_Bears.rst [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+We can test that by giving it a sample JSON string. Before testing we need to
+build it though. The JSON library requires C++11 so a sample ``Makefile`` would
+look like this:
+
+::
+
+    build: coalaCheck.cpp
+        g++ -std=c++11 -o coalaCheck_cpp coalaCheck.cpp
+
+We test our binary and it work but we don't quite do any code analysis yet.
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: work -> works

File: docs/Users/Tutorials/External_Bears.rst [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+And we are pretty much done. In order to run this Bear we do not need to add the
+source code to the path because the binary being run is ``node``. Although there
+is a problem: the argument we supplied will be looked up only in the current
+directory. To fix this we can add the full path of the ``.js`` file in the
+argument list. In this case we just run the bear from the same directory as
+``coalaCheck.js``. The code for this example can be found
+`here <https://github.com/Redridge/coalaCheckBear-js>`__.
+
+The JSON Spec
```
Comment (arafsheikh) [inline docs/Users/Tutorials/External_Bears.rst]: It would be better to move this section to a new page in the docs.

File: docs/Users/Tutorials/External_Bears.rst [hunk 8]
```diff
@@ -0,0 +1,511 @@
+External Bears
+==============
+
+Welcome. This tutorial hopes to teach you how to use the
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/External_Bears.rst]: cmon you cant use in a tutorial ``hopes to teach you`` :p

try: This tutorial is made to teach you..

File: docs/Users/Tutorials/External_Bears.rst [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+``@external_bear_wrap`` decorator in order to write Bears in languages other
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/External_Bears.rst]: no need for capital B imo

File: docs/Users/Tutorials/External_Bears.rst [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+than python.
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/External_Bears.rst]: here should be Python

File: docs/Users/Tutorials/External_Bears.rst [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+
+.. note::
+
+  This tutorial assumes that you already know the basics. If you are new please
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/External_Bears.rst]: comma after new

File: docs/Users/Tutorials/External_Bears.rst [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+
+In this tutorial we will go through 2 examples where we create a very simple
+bear. The first example will use a **compiled** language that creates a
+standalone binary, ``C++``, whilst in the second example we will take a look at
+``JS`` that needs ``node`` in order to run out of the browser.
+
+External Bear Generation Tool
+-----------------------------
+
+If you really do not want to write any python code, there is a tool
```
Comment (Adrianzatreanu) [inline docs/Users/Tutorials/External_Bears.rst]: Python

File: docs/Users/Tutorials/External_Bears.rst (hunk 8 above)
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: "This tutorial will teach you....." is better.

File: docs/Users/Tutorials/External_Bears.rst [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+
+::
+
+    $ coala-bears-create -ext
+
+in order to generate the wrapper for the bear.
+
+Writing a Bear in C++
+---------------------
+
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Make this in steps.

Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Don't use we will. Make points like:
* Create a new directory.

File: docs/Users/Tutorials/External_Bears.rst (hunk 4 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: I am trying to explain why it is useful to have the feature of writing bears in other languages.
Someone doing research in code analysis might already be proficient in java and might not want to use python

File: docs/Users/Tutorials/External_Bears.rst (hunk 9 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: I would rather use B

File: docs/Users/Tutorials/External_Bears.rst (hunk 7 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: It is used only for external bears though
I put it here and use links to it (see C++ tutorial)

File: docs/Users/Tutorials/External_Bears.rst [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+Enabling developers to write external Bears means that they do not need to waste
+any time learning new languages and they can get faster to the interesting part,
+**static code analysis**.
+
+How Does This Work?
+-------------------
+
+By using the ``@external_bear_wrap`` decorator you will have all the necessary
+data sent to your external executable (filename, lines, settings) as a JSON
+string via ``stdin``. Afterwards the analysis takes place in your executable
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Afterwards,

File: docs/Users/Tutorials/External_Bears.rst [hunk 15]
```diff
@@ -0,0 +1,521 @@
...
+that can be written in **literally** any language. In the end you will have to
+provide the ``Results`` in a JSON string via stdout.
+
+In this tutorial we will go through 2 examples where we create a very simple
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: tutorial,

File: docs/Users/Tutorials/External_Bears.rst [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+any time learning new languages and they can get faster to the interesting part,
+**static code analysis**.
+
+How Does This Work?
+-------------------
+
+By using the ``@external_bear_wrap`` decorator you will have all the necessary
+data sent to your external executable (filename, lines, settings) as a JSON
+string via ``stdin``. Afterwards the analysis takes place in your executable
+that can be written in **literally** any language. In the end you will have to
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: In the end,

File: docs/Users/Tutorials/External_Bears.rst [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+::
+
+    #include <iostream>
+
+    #include "json.hpp"
+
+    using json = nlohmann::json;
+    using namespace std;
+
+    json in;
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Can we rename this as input?

File: docs/Users/Tutorials/External_Bears.rst [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+string via ``stdin``. Afterwards the analysis takes place in your executable
+that can be written in **literally** any language. In the end you will have to
+provide the ``Results`` in a JSON string via stdout.
+
+In this tutorial we will go through 2 examples where we create a very simple
+bear. The first example will use a **compiled** language that creates a
+standalone binary, ``C++``, whilst in the second example we will take a look at
+``JS`` that needs ``node`` in order to run out of the browser.
+
+External Bear Generation Tool
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: This section here breaks the flow of the tutorial. Put this at the end. Before the json spec.

Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Oh ok nvm.

File: docs/Users/Tutorials/External_Bears.rst [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+to make.
+
+1. Create a new directory and make it your current working directory.
+2. Run ``coala-bears-create`` as mentioned above in order to create the wrapper
+   for our ``C++`` bear and hit ``Enter`` for the first question. You can skip
+   this step and use the wrapper provided below.
+3. Name the bear ``coalaCheckBear`` and the executable ``coalaCheck_cpp``.
+4. Reply to the rest of the questions (hit ``Enter`` at settings). After the
+   script is finished running there should be 2 files in your current directory:
+   ``coalaCheckBear.py`` (the wrapper) and ``coalaCheckBearTest.py``.
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: What do they need to reply. Do they simply have to go with the defaults? Because I don't think first time bear writers would understand this well.

Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Just very briefly explain every question asked.

File: docs/Users/Tutorials/External_Bears.rst [hunk 20]
```diff
@@ -0,0 +1,521 @@
...
+5. This tutorial will not focus on testing so ignore the second file for now.
+   The wrapper should look something like this (after some cleaning):
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: The wrapper should look similar to the code presented below. Some code has been cleaned for convenience of explanation.

^Is that what you want to say?

File: docs/Users/Tutorials/External_Bears.rst [hunk 21]
```diff
@@ -0,0 +1,521 @@
...
+        AUTHORS = {'Me'}
+        AUTHORS_EMAILS = {'me@mail.com'}
+        LICENSE = 'GPLv3'
+
+        @staticmethod
+        def create_arguments():
+            return ()
+
+6. Since the input will be a JSON string some kind of JSON class is needed.
+   `This one <https://github.com/nlohmann/json>`__ is a great choice because it
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Just give the full link because you refer to the json library as nlohmann's JSON library.

File: docs/Users/Tutorials/External_Bears.rst [hunk 22]
```diff
@@ -0,0 +1,521 @@
...
+    #include <iostream>
+    #include <string>
+
+    #include "json.hpp"
+
+    using json = nlohmann::json;
+    using namespace std;
+
+    json in;
+    json out;
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: just to keep uniformity make this output.

File: docs/Users/Tutorials/External_Bears.rst [hunk 23]
```diff
@@ -0,0 +1,521 @@
...
+
+        cin >> in;
+
+        init_results("coalaCheckBear");
+
+        cout << out;
+        return 0;
+    }
+
+The ``C++`` operators and syntax are not well suited for JSON manipulation but
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: This is out of indentation. Either make it a Note or make it part of the above point and code with correct indentation.

File: docs/Users/Tutorials/External_Bears.rst [hunk 24]
```diff
@@ -0,0 +1,521 @@
...
+            }
+        }
+
+        cout << out;
+
+        return 0;
+    }
+
+After building the executable it has to be added it to the ``PATH`` env
+variable. It is possible to modify the wrapper and give there the full path.
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Same problem here make it part of a bullet point.

File: docs/Users/Tutorials/External_Bears.rst (hunk 17 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: I think I didn't do it because sublime colored it differently, hinting that it is a key word.

File: docs/Users/Tutorials/External_Bears.rst (hunk 20 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: You took the words from my mouth :)

File: docs/Users/Tutorials/External_Bears.rst (hunk 23 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: I wanted it to be there. I will go with the note. Initially I didn't want to because too many notes may be annoying (not sure)

File: docs/Users/Tutorials/External_Bears.rst [hunk 25]
```diff
@@ -0,0 +1,529 @@
...
+  linter), please refer to the
+  :doc:`Linter Bears<Linter_Bears>` section.
+
+Why is This Useful?
+-------------------
+
+coala is a great tool if you want to write static code analysis without having
+to worry about the interaction with the user. That makes it an excellent choice
+for researchers and other people who want to build new static analysis routines.
+Enabling developers to write external Bears means that they do not need to waste
```
Comment (SanketDG) [inline docs/Users/Tutorials/External_Bears.rst]: Put a newline before this line so that the reason and solution can be distinguished. Also I would trim down on the reasoning a little. Something like:

```
coala is great language independent static analysis tool, where users can write their own static analysis routines. Enabling users to write external ``bears`` means that they can write their own static analysis routine in their favourite language.
```

File: docs/Users/Tutorials/External_Bears.rst [hunk 26]
```diff
@@ -0,0 +1,529 @@
...
+
+By using the ``@external_bear_wrap`` decorator you will have all the necessary
+data sent to your external executable (filename, lines, settings) as a JSON
+string via ``stdin``. Afterwards, the analysis takes place in your executable
+that can be written in **literally** any language. In the end, you will have to
+provide the ``Results`` in a JSON string via stdout.
+
+In this tutorial, we will go through 2 examples where we create a very simple
+bear. The first example will use a **compiled** language that creates a
+standalone binary, ``C++``, whilst in the second example we will take a look at
```
Comment (SanketDG) [inline docs/Users/Tutorials/External_Bears.rst]:
```
The first example will use a **compiled** language, ``C++``, that creates a standalone binary, whilst...
````

File: docs/Users/Tutorials/External_Bears.rst [hunk 27]
```diff
@@ -0,0 +1,529 @@
...
+    variable. It is possible to modify the wrapper and give there the full path.
+    Add the current directory to the ``PATH`` like so:
+
+::
+
+    $ export PATH=$PATH:$PWD
+
+The last step is to test if everything is working properly. This is the testfile
+used in this tutorial (
+`testfile <https://github.com/Redridge/coalaCheckBear-cpp>`__).
```
Comment (SanketDG) [inline docs/Users/Tutorials/External_Bears.rst]: did you mean the link to be https://raw.githubusercontent.com/Redridge/coalaCheckBear-cpp/master/testfile?

Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: yes indeed

**Comment by sils1297** [comment]
there's already a review, could you mark WIP in the future @SanketDG ?

File: docs/Users/Tutorials/External_Bears.rst [hunk 28]
```diff
@@ -0,0 +1,528 @@
...
+
+
+    @external_bear_wrap(executable='coalaCheck_cpp',
+                        settings={})
+    class coalaCheckBear:
+        """
+        Checks for coala written with uppercase 'C'
+        """
+        LANGUAGES = {'All'}
+        REQUIREMENTS = {'coalaCheck_cpp'}
```
Comment (sils1297) [inline docs/Users/Tutorials/External_Bears.rst]: that's not a valid requirement, huh? CC @Adrianzatreanu

File: docs/Users/Tutorials/External_Bears.rst [hunk 29]
```diff
@@ -0,0 +1,528 @@
...
+        AUTHORS = {'Me'}
+        AUTHORS_EMAILS = {'me@mail.com'}
+        LICENSE = 'GPLv3'
```
Comment (sils1297) [inline docs/Users/Tutorials/External_Bears.rst]: can we have a comment generated here that the license applies only to e.g the C++ code; the python code will legally be AGPL

File: docs/Users/Tutorials/External_Bears.rst [hunk 30]
```diff
@@ -0,0 +1,528 @@
...
+
+::
+
+    build: coalaCheck.cpp
+        g++ -std=c++11 -o coalaCheck_cpp coalaCheck.cpp
+
+9. Compile and test the binary by giving it a JSON string. It should print the
+   JSON string back at ``stdout``.
+
+10. Read about the JSON Spec that the input uses(`The JSON Spec`_). The filename
```
Comment (sils1297) [inline docs/Users/Tutorials/External_Bears.rst]: what kind of weird sntax is this? Don't we want a space after uses?

Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: Yeah, it's a typo, sorry

File: docs/Users/Tutorials/External_Bears.rst (hunk 29 above)
Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: Oh, no, that applies to the python code. The `C++` license should be in the `C++` file. I will swap it to AGPL to match the rest.

File: docs/Users/Tutorials/External_Bears.rst [hunk 31]
```diff
@@ -0,0 +1,528 @@
...
+                        settings={})
+    class coalaCheckBear:
+        """
+        Checks for coala written with uppercase 'C'
+        """
+        LANGUAGES = {'All'}
+        REQUIREMENTS = {''}
+        AUTHORS = {'Me'}
+        AUTHORS_EMAILS = {'me@mail.com'}
+        LICENSE = 'AGPL'
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Okay so just mention somewhere about how the C++ file will have a separate license. or better yet add a license to your C++ code as well to make it explicitly clear to the reader about the situation.

Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: Adding a license to the C++ source code is irrelevant to the tutorial. I will just make a mention about that.

File: docs/Users/Tutorials/External_Bears.rst [hunk 32]
```diff
@@ -0,0 +1,533 @@
...
+   script is finished running there should be 2 files in your current directory:
+   ``coalaCheckBear.py`` (the wrapper) and ``coalaCheckBearTest.py``.
+5. This tutorial will not focus on testing so ignore the second file for now.
+   The wrapper should look similar to the code block presented below. Some code
+   has been cleaned for convenience of explanation.
+
+.. note::
+
+    The ``LICENSE`` specified applies only to the python code. You can license
+    you executable however you see fit.
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: typo: your

File: docs/Users/Tutorials/External_Bears.rst [hunk 33]
```diff
@@ -0,0 +1,533 @@
...
+                        settings={})
+    class coalaCheckBear:
+        """
+        Checks for coala written with uppercase 'C'
+        """
+        LANGUAGES = {'All'}
+        REQUIREMENTS = {'node'}
+        AUTHORS = {'Me'}
+        AUTHORS_EMAILS = {'me@mail.com'}
+        LICENSE = 'GPLv3'
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: This should be AGPL

File: docs/Users/Tutorials/External_Bears.rst [hunk 34]
```diff
@@ -0,0 +1,533 @@
...
+
+When another binary is needed to run the source code, the ``create_arguments``
+method comes in handy.
+
+2. Add the source code file as an argument to the ``create_arguments`` method
+   (so that the command becomes ``node coalaCheck.js``).
+
+The ``create_arguments`` method returns a tuple so if only one
+argument is added then a comma has to be used at the end (e.g. ``(one_item,)``).
+
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: Mention the license here too.

File: docs/Users/Tutorials/External_Bears.rst [hunk 35]
```diff
@@ -0,0 +1,538 @@
...
+provide a JSON string via stdout. The specs are the following:
+
+* input JSON spec
+
++--------------------------------+-------+-----------------------------------+
+|Tree                            |Type   |Description                        |
++--------------------------------+-------+-----------------------------------+
+|filename                        |str    |the name of the file being analysed|
++--------------------------------+-------+-----------------------------------+
+|file                            |list   |file contents as a list of files   |
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: File contents as list of files.... It is not clear what you are trying to say?

Comment (Redridge) [inline docs/Users/Tutorials/External_Bears.rst]: the file contents will be passed as a list of files

Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: You mean a list of lines. What is list of files?

File: docs/Users/Tutorials/External_Bears.rst [hunk 36]
```diff
@@ -0,0 +1,538 @@
...
+provide a JSON string via stdout. The specs are the following:
+
+* input JSON spec
+
++--------------------------------+-------+-----------------------------------+
+|Tree                            |Type   |Description                        |
++--------------------------------+-------+-----------------------------------+
+|filename                        |str    |the name of the file being analysed|
++--------------------------------+-------+-----------------------------------+
+|file                            |list   |file contents as a list of liles   |
```
Comment (Udayan12167) [inline docs/Users/Tutorials/External_Bears.rst]: typo: lines :P

**Comment by Udayan12167** [comment]
ack 03fcb55

**Comment by Udayan12167** [comment]
@rultor merge
